File: solution/238ProductOfArrayExceptSelf/238ProductOfArrayExceptSelf.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  vector<int> productExceptSelf1(vector<int> &nums) {
    int n = nums.size();
    vector<int> L(n, 1);
    vector<int> R(n, 1);
    for (int i = 1; i < n; i++) {
      L[i] = L[i - 1] * nums[i - 1];
    }
    for (int i = n - 2; i >= 0; i--) {
      R[i] = R[i + 1] * nums[i + 1];
    }
    vector<int> result(n);
    for (int i = 0; i < n; i++) {
      result[i] = L[i] * R[i];
    }
    return result;
  }

  vector<int> productExceptSelf(vector<int> &nums) {
    int n = nums.size();
    vector<int> result(n, 1);
    for (int i = 1; i < n; i++) {
      result[i] = result[i - 1] * nums[i - 1];
    }
    int R = 1;
    for (int i = n - 2; i >= 0; i--) {
      R = R * nums[i + 1];
      result[i] *= R;
    }
    return result;
  }
};

int main() {}

File: src/238ProductOfArrayExceptSelf/238ProductOfArrayExceptSelf.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len(); 
        assert!(n>1);
        
        let mut left  = vec![1; n];
        let mut right = vec![1; n];
        for i in 1..n {
            left[i] = left[i-1]*nums[i-1];
            right[n-1-i] = right[n-i]*nums[n-i];
        }
        let mut result = vec![0; n];
        for i in 0..n {
            result[i] = left[i]*right[i];
        }
        return result;
    }
}
